src/app/interaction-model/InvokeResponder.h

```diff
@@ -0,0 +1,227 @@
+    ///
+    virtual CHIP_ERROR Complete(CHIP_ERROR error) = 0;
+};
+
+/// Enforces that once acquired, Complete will be called on the underlying writer
+class AutoCompleteInvokeResponder
+{
+public:
+    // non-copyable: once you have a handle, keep it
+    AutoCompleteInvokeResponder(const AutoCompleteInvokeResponder &)             = delete;
+    AutoCompleteInvokeResponder & operator=(const AutoCompleteInvokeResponder &) = delete;
+
+    AutoCompleteInvokeResponder(InvokeResponder * writer) : mWriter(writer) {}
+    ~AutoCompleteInvokeResponder()
+    {
+        if (!mCompleted)
+        {
+            mWriter->Complete(CHIP_IM_GLOBAL_STATUS(Failure));
+        }
+    }
+
+    /// Direct access to reply encoding.
+    ///
+    /// Use this only in conjunction with the other Raw* calls
+    DataModel::WrappedStructEncoder & RawResponseEncoder(CommandId replyCommandId)
+    {
+        return mWriter->ResponseEncoder(replyCommandId);
+    }
+
+    /// Direct access to flushing replies
+    ///
+    /// Use this only in conjunction with the other Raw* calls
+    CHIP_ERROR RawFlushPendingReplies()
+    {
+        mCompleted = false;
+        return mWriter->FlushPendingResponses();
+    }
+
+    /// Call "Complete" without the automatic retries.
+    ///
+    /// Use this in conjunction with the other Raw* calls
+    CHIP_ERROR RawComplete(CHIP_ERROR error)
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted = true;
+        return mWriter->Complete(error);
+    }
+
+    /// Complete the given command.
+    ///
+    /// Automatically handles retries for sending.
+    ///
+    /// Any error returned by this are final and not retriable
+    /// as a retry for CHIP_ERROR_BUFFER_TOO_SMALL is already built in.
+    CHIP_ERROR Complete(CHIP_ERROR error)
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted     = true;
+        CHIP_ERROR err = mWriter->Complete(error);
+
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return err;
+        }
+
+        // retry once. Failure to flush is permanent.
+        ReturnErrorOnFailure(mWriter->FlushPendingResponses());
+        return mWriter->Complete(error);
+    }
+
+    /// Sends the specified data structure as a response
+    ///
+    /// This version of the send has built-in RETRY and handles
+    /// Flush/Complete automatically.
+    ///
+    /// Any error returned by this are final and not retriable
+    /// as a retry for CHIP_ERROR_BUFFER_TOO_SMALL is already built in.
+    template <typename ReplyData>
+    CHIP_ERROR Send(const ReplyData & data)
+    {
+        VerifyOrReturnError(!mCompleted, CHIP_ERROR_INCORRECT_STATE);
+        mCompleted     = true;
+        CHIP_ERROR err = data.Encode(ResponseEncoder(ReplyData::GetCommandId()));
+        if (err != CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            LogErrorOnFailure(err);
+            err = mWriter->Complete(err);
```

Comment:
If `FlushPendingReplies` is responsible for rolling back the buffer (as per the comment [here](https://github.com/project-chip/connectedhomeip/pull/32914/files#r1567540325)) what are we relying on here to rollback so that the `err` can be added here?

Reply:
My expectation is that the writer is responsible to have a valid state, including rolling back data from commands in order to be able to encode an error. So in this case, `Encode` return an error so the writer is expected to send a status error.

The updated comments in the `Complete` of writers hopefully explain this more clearly.